OCD-4147: Show attestation warnings for ONC-ACB users

kekey1 opened the pull request.

kekey1 committed:
refactor: Move RWT classes into RWT-specific package

[#OCD-4147]

kekey1 committed:
refactor: Pull attestation validation logic out of report class

We want to be able to validate attestations in places besides generating
the report, so I pulled ou some of the checking code into a couple of
services and am just using that code in the report class now.

[#OCD-4147]

kekey1 committed:
fix: Remove state saved in the CheckInReportDataCollector bean

The developerListings attribute was saving a map of developer IDs to
that developer's active listings. If the following steps were followed,
there was the possibility of putting an incorrect warning message in the
report:
1) Run the Check-In Report
	a) This saves all the active listings for every developer and
	uses that set of listings to insert warning messages into the
	report
2) Edit an active listing for one of the developers in a way that you
   would expect it's warning message to change.
   	ex: If the developer had answered Compliant for their API
	Response but there was a warning that none of their listings had
	g7-g10 criteria (so they should have answered Not Applicable),
	and you edit an active listing for that developer to have it
	attest to g7.
3) Run the Check-In Report again
	a) The same CheckInReportDataCollector bean is re-used and it
	references it's in-memory set of active listings for each
	developer, even the one you just edited, so the warning remains
	in the report even though it no longer should be there.

[#OCD-4147]

kekey1 committed:
refactor: Slightly improve performance of the check-in report

Gathering the information about the most recent attestation change
request for the current attestation period for each developer was
performing linearly and would increase over time each quarter when more
change requests are added for each developer. Locally it was taking
about 3 seconds per developer and around 20 minutes total to run the
report and would slowly increase over time. This update runs in about
15 minutes locally and I don't *believe* it will increase over time.

The previous code was getting all attestation change requests for a
developer, then getting the details of each of those change requests in
order to sort them by submission date. The second step would be
increasing in time as more attestation periods go by. I changed this to
make a query to get the ID of the most recent attestation change request
for the developer and then a second query to get the details of that
change request. I think in the long term this will perform better.

[#OCD-4147]

kekey1 committed:
feat!: Warn ACBs about inconsistent attestations responses

[#OCD-4147]

kekey1 committed:
feat!: Require acknowledgement of warnings on a change request

[#OCD-4147]

kekey1 committed:
feat-flag!: Ignore most assurances warnings until erd-phase-3

If the developer has a listing with b10 but gives an Assurances
attestation that they are compliant because they do not have listings
with b10, we will give that warning always. Any other Assurances or b10
related warnings will not be given until the erd-phase-3 flag is on.

[#OCD-4147]

kekey1 committed:
Merge remote-tracking branch 'upstream/staging' into OCD-4147

kekey1 committed:
feat: Update attestation warning text to match ONC-provided text

[#OCD-4147]

kekey1 committed:
fix: cr unit test compilation issues

[#OCD-4147]

kekey1 committed:
test: Fix broken change request permissions tests

[#OCD-4147]

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
feat: Add validation for attestation period

Ensure the attestation period passed-in on a POST or PUT to
change-requests is the same as the currently submittable period for the
developer.

[#OCD-4147]

tmy1313 commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/service/ListingApplicabilityService.java:
I like this class, but think maybe it should be renamed or moved to a different package.  Since this is not where we globally determine if a listing is RWT applicable, someone might think this how to determine if a listing is eligible for RWT.  I would suggest adding the word "attestation" to the class name or moving it to an attestation specific package

kekey1 replied:
Moved! Thank you for the feedback and you are absolutely right.

kekey1 committed:
refactor: Moved service into attestation-specific package

The ListingApplicabilityService is currently only relevant in the
context of attestations so it was moved to the attestation.service
package to reflect that.

[#OCD-4147]

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
Merge branch 'staging' into OCD-4147

kekey1 committed:
style: Correct newlining around boolean operator

[#OCD-4147]
